### PR TITLE
Set shard mode to read-only if there are too many errors

### LINKED ISSUE #1118: Make shard read-only on errors
As described in https://github.com/nspcc-dev/neofs-node/issues/1094#issuecomment-1018259445, there are many places where unexpected error is encountered during shard operation. We can make shard read-only in this situation, carefully checking that this is indeed a reasonable action for all possible errors returned.

Related #1115 .

### PR DESCRIPTION
**Pull request by fyrchik**
Close #1118 .

Error counter can be cleared with an additional field via `SetShardMode` and is reported in `ListShards` information.

### CODE REVIEW

File: config/example/node.yaml [hunk 1]
```diff
@@ -99,6 +99,7 @@ storage:
   # note: shard configuration can be omitted for relay node (see `node.relay`)
   shard_pool_size: 15 # size of per-shard worker pools used for PUT operations
   shard_num: 2  # total number of shards
+  shard_ro_error_threshold: 100
```
Comment (cthulhu-rider) [inline config/example/node.yaml:102]: Name is quite telling, but let's still add a verbal explanation (in particular, that with a zero/NaN value, the restriction is not set)

Comment (fyrchik) [reply config/example/node.yaml]: Fixed.

**Review by carpawell — APPROVED** [review]
the first commit message body contains the typo in the `during` word

also, small style notes: we try to write commit message in the following form:
`[#xxxx] zone: Capitalized_word`

File: pkg/local_object_storage/engine/shards.go [hunk 2]
```diff
@@ -91,23 +93,23 @@ func (e *StorageEngine) unsortedShards() []hashedShard {
 	shards := make([]hashedShard, 0, len(e.shards))
 
 	for _, sh := range e.shards {
-		shards = append(shards, hashedShard{sh})
+		shards = append(shards, hashedShard{errorCount: sh.errorCount, Shard: sh.Shard})
```
Comment (carpawell) [inline pkg/local_object_storage/engine/shards.go:96]: linter

```suggestion
		shards = append(shards, hashedShard(sh))
```

File: pkg/local_object_storage/engine/engine.go [hunk 3]
```diff
@@ -26,12 +27,49 @@ type StorageEngine struct {
...
+	*shard.Shard
+}
+
+// reportShardError checks that amount of errors doesn't exceed configured threshold.
+// If it does, shard is set to read-only mode.
+func (e *StorageEngine) reportShardError(
+	sh hashedShard,
+	msg string,
+	err error,
+	fields ...zap.Field) {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/local_object_storage/engine/engine.go:41]: The line would be too long.

File: pkg/local_object_storage/engine/shards.go (hunk 2 above)
Comment (fyrchik) [reply pkg/local_object_storage/engine/shards.go:96]: Fixed.

File: pkg/local_object_storage/engine/engine.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/local_object_storage/engine/engine.go:41]: just to know: what length is suitable for you? i use 120 char threshold that is why this comment appeared

File: cmd/neofs-node/config/engine/config.go [hunk 4]
```diff
@@ -58,3 +58,10 @@ func ShardPoolSize(c *config.Config) uint32 {
 
 	return ShardPoolSizeDefault
 }
+
+// ShardErrorThreshold returns value of "shard_ro_error_threshold" config parameter from "storage" section.
+//
+// Returns 0 if the value is missing.
```
Comment (alexvanin) [inline cmd/neofs-node/config/engine/config.go:64]: I think it is not good to be fault intolerant by default, but I don't have any good default value for that, 10, 100, who knows...

File: pkg/local_object_storage/engine/engine.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/local_object_storage/engine/engine.go:41]: Lately I try to wrap lines at 100.

**Review by alexvanin — APPROVED** [review]
LGTM. While global error counter is okay, I also thought more about sampler.

Sampler resets error counter every `t` seconds and stores up to `n` recent samples. If sum of errors in `n` samples > threshold: enable read-only mode.

File: cmd/neofs-node/config/engine/config.go (hunk 4 above)
Comment (fyrchik) [reply cmd/neofs-node/config/engine/config.go:64]: I think this non-zero default is good, but it should be provided in config, shard should not define any policy by itself IMO.

Comment (alexvanin) [reply cmd/neofs-node/config/engine/config.go:64]: `neofs-node/config` package was designed for default app configuration, so it fits there. But still don't know what would be a good value for that.

**Pull request merged**